### PR TITLE
Install signal handler to aid debugging

### PR DESCRIPTION
**Pull request by athos-ribeiro**
Here is a first proposal for installing a signal handler in Atomic Reactor.
I do have questions:

1.  Would it be preferable to log the messages (as in the proposed commit), or dump them somewhere else, namely, `STDOUT`?
2. Should we always leave the listener on or should the feature only be available in a debugging mode?

### CODE REVIEW

**Comment by twaugh** [comment]
Nice work!

We should probably attempt to show tracebacks for all threads (e.g. with `sys._current_frames`). Perhaps we can use `faulthandler` if on Python 3?

I think we should write the output directly to `sys.stdout` or even `sys.stderr`.

We should always have it enabled, because we occasionally see issues in production that a stack trace would help us to diagnose.

**Comment by MartinBasti** [comment]
I quite don't like that the signal handling is scattered in various modules, something is in `inner.py`, something is in `utils.py`. Would to possible to create module with all signal handlers defined in one place, and maybe nicely named to be know from name which ones are for builds only, which for whole application?

[optional, just my 2 cents]

File: atomic_reactor/util.py [hunk 1]
```diff
@@ -1530,3 +1532,11 @@ def make_list(value):
                 )
             expected_platforms = expected_platforms & only_platforms
     return expected_platforms - excluded_platforms
+
+
+def dump_stacktrace(sig, frame):
+    logger.info("Signal for traceback received:\n%s", ''.join(traceback.format_stack(frame)))
+
+
+def start_signal_handler():
```
Comment (lcarva) [inline atomic_reactor/util.py]: Maybe call this something more specific? `setup_introspection_signal_handler` ?

**Comment by athos-ribeiro** [comment]
Sample output snippet for the orchestrator:

```
2018-10-24 08:46:31,276 platform:- - osbs.conf - WARNING - it has been deprecated in favor of the value in the reactor_config_map
2018-10-24 08:46:31,276 platform:- - osbs.build.user_params - WARNING - build_image or build_imagestream is defined, they are deprecated,use build_from instead
2018-10-24 08:46:31,349 platform:- - atomic_reactor.plugins.orchestrate_build - INFO - x86_64 - created build hello-osbs-master-f6a0a-1 on cluster worker.
Thread 0x7f566e7fc700 (most recent call first):
  File "/usr/lib64/python2.7/threading.py", line 339, in wait
  File "/usr/lib64/python2.7/Queue.py", line 168, in get
  File "/usr/lib64/python2.7/multiprocessing/pool.py", line 376, in _handle_results
  File "/usr/lib64/python2.7/threading.py", line 765, in run
  File "/usr/lib64/python2.7/threading.py", line 812, in __bootstrap_inner
  File "/usr/lib64/python2.7/threading.py", line 785, in __bootstrap
Thread 0x7f566effd700 (most recent call first):
  File "/usr/lib64/python2.7/threading.py", line 339, in wait
  File "/usr/lib64/python2.7/Queue.py", line 168, in get
  File "/usr/lib64/python2.7/multiprocessing/pool.py", line 335, in _handle_tasks
  File "/usr/lib64/python2.7/threading.py", line 765, in run
  File "/usr/lib64/python2.7/threading.py", line 812, in __bootstrap_inner
  File "/usr/lib64/python2.7/threading.py", line 785, in __bootstrap
Thread 0x7f566f7fe700 (most recent call first):
  File "/usr/lib64/python2.7/multiprocessing/pool.py", line 326, in _handle_workers
  File "/usr/lib64/python2.7/threading.py", line 765, in run
  File "/usr/lib64/python2.7/threading.py", line 812, in __bootstrap_inner
  File "/usr/lib64/python2.7/threading.py", line 785, in __bootstrap
Current thread 0x7f5682285740 (most recent call first):
  File "/usr/lib/python2.7/site-packages/atomic_reactor/util.py", line 74, in dump_traceback
  File "/usr/lib/python2.7/site-packages/atomic_reactor/util.py", line 1562, in dump_stacktraces
  File "/usr/lib64/python2.7/threading.py", line 339, in wait
  File "/usr/lib64/python2.7/multiprocessing/pool.py", line 543, in wait
  File "/usr/lib64/python2.7/multiprocessing/pool.py", line 548, in get
  File "/usr/lib/python2.7/site-packages/atomic_reactor/plugins/build_orchestrate_build.py", line 861, in run
  File "/usr/lib/python2.7/site-packages/atomic_reactor/plugin.py", line 239, in run
  File "/usr/lib/python2.7/site-packages/atomic_reactor/plugin.py", line 449, in run
  File "/usr/lib/python2.7/site-packages/atomic_reactor/inner.py", line 443, in build_docker_image
  File "/usr/lib/python2.7/site-packages/atomic_reactor/inner.py", line 546, in build_inside
  File "/usr/lib/python2.7/site-packages/atomic_reactor/cli/main.py", line 95, in cli_inside_build
  File "/usr/lib/python2.7/site-packages/atomic_reactor/cli/main.py", line 292, in run
  File "/usr/lib/python2.7/site-packages/atomic_reactor/cli/main.py", line 310, in run
  File "/usr/bin/atomic-reactor", line 11, in <module>
Thread 0x7f5674ee1700 (most recent call first):
  File "/usr/lib/python2.7/site-packages/atomic_reactor/inner.py", line 277, in run
  File "/usr/lib64/python2.7/threading.py", line 812, in __bootstrap_inner
  File "/usr/lib64/python2.7/threading.py", line 785, in __bootstrap
Thread 0x7f566ffff700 (most recent call first):
  File "/usr/lib64/python2.7/threading.py", line 339, in wait
  File "/usr/lib64/python2.7/Queue.py", line 168, in get
  File "/usr/lib64/python2.7/multiprocessing/pool.py", line 102, in worker
  File "/usr/lib64/python2.7/threading.py", line 765, in run
  File "/usr/lib64/python2.7/threading.py", line 812, in __bootstrap_inner
  File "/usr/lib64/python2.7/threading.py", line 785, in __bootstrap
2018-10-24 08:52:09,606 platform:- - atomic_reactor.plugins.orchestrate_build - DEBUG - all koji-build-ids: set([None])
2018-10-24 08:52:09,606 platform:- - atomic_reactor.plugin - DEBUG - plugin 'orchestrate_build' finished in 338s
2018-10-24 08:52:09,606 platform:- - atomic_reactor.plugin - DEBUG - stopping further execution of plugins after first successful plugin
2018-10-24 08:52:09,606 platform:- - atomic_reactor.plugin - INFO - initializing runner of pre-publish plugins
```

Sample output snippet for build:

```
018-10-24 08:46:33,705 - atomic_reactor.plugin - DEBUG - load file '/usr/lib/python2.7/site-packages/atomic_reactor/plugins/prepub_squash.py'
2018-10-24 08:46:33,705 - atomic_reactor.plugin - DEBUG - load file '/usr/lib/python2.7/site-packages/atomic_reactor/plugins/prepub_tests_for_image.py'
2018-10-24 08:46:33,705 - atomic_reactor.plugin - INFO - building image ImageName(image=u'kojiadmin/hello-osbs:candidate-64139-20181024084631-x86_64') inside current environment
2018-10-24 08:46:33,705 - atomic_reactor.plugin - DEBUG - using dockerfile:
FROM sha256:f6f9d2ff8a74402d428ced8539338d5082d069c9d6982cf8d8fb9efed759008d

LABEL "com.redhat.component"="hello-osbs" \
      "name"="athos-ribeiro/hello-osbs" \
      "version"="1.0"

RUN sleep 300

ADD Dockerfile-athos-ribeiro-hello-osbs-1.0-1 /root/buildinfo/Dockerfile-athos-ribeiro-hello-osbs-1.0-1
LABEL "authoritative-source-url"="osbs-box.example.com" "distribution-scope"="public" "vendor"="Red Hat, Inc." "vcs-type"="git" "architecture"="x86_64" "release"="1" "build-date"="2018-10-24T08:46:32.133885" "com.redhat.build-host"="athos-osbs-box" "vcs-ref"="ea134957a9781eb3dad8f6654732f4e84ff1a51d"

2018-10-24 08:46:33,705 - atomic_reactor.plugin - DEBUG - running plugin 'docker_api'
2018-10-24 08:46:33,705 - atomic_reactor.plugin - INFO - running plugin instance with args: '{}'
2018-10-24 08:46:33,705 - atomic_reactor.core - INFO - building image 'kojiadmin/hello-osbs:candidate-64139-20181024084631-x86_64' from path '/tmp/tmpEXi_UN/hello-osbs'
2018-10-24 08:46:33,774 - atomic_reactor.plugins.docker_api - DEBUG - build is submitted, waiting for it to finish
2018-10-24 08:46:33,774 - atomic_reactor.util - INFO - wait_for_command
2018-10-24 08:46:33,774 - atomic_reactor.util - DEBUG - Step 1/5 : FROM sha256:f6f9d2ff8a74402d428ced8539338d5082d069c9d6982cf8d8fb9efed759008d
2018-10-24 08:46:33,774 - atomic_reactor.util - DEBUG - ---> f6f9d2ff8a74
2018-10-24 08:46:33,775 - atomic_reactor.util - DEBUG - Step 2/5 : LABEL "com.redhat.component" "hello-osbs" "name" "athos-ribeiro/hello-osbs" "version" "1.0"
2018-10-24 08:46:33,831 - atomic_reactor.util - DEBUG - ---> Running in ad44e540ad54
2018-10-24 08:46:34,000 - atomic_reactor.util - DEBUG - ---> aa50ddb0bfa6
2018-10-24 08:46:34,011 - atomic_reactor.util - DEBUG - Removing intermediate container ad44e540ad54
2018-10-24 08:46:34,012 - atomic_reactor.util - DEBUG - Step 3/5 : RUN sleep 300
2018-10-24 08:46:34,060 - atomic_reactor.util - DEBUG - ---> Running in 55065d909a13
2018-10-24 08:46:34,061 - atomic_reactor.util - DEBUG - [91m[0m
Thread 0x7f071c0ad700 (most recent call first):
  File "/usr/lib/python2.7/site-packages/atomic_reactor/inner.py", line 277, in run
  File "/usr/lib64/python2.7/threading.py", line 812, in __bootstrap_inner
  File "/usr/lib64/python2.7/threading.py", line 785, in __bootstrap

Current thread 0x7f0729451740 (most recent call first):
  File "/usr/lib/python2.7/site-packages/atomic_reactor/util.py", line 74, in dump_traceback
  File "/usr/lib/python2.7/site-packages/atomic_reactor/util.py", line 1562, in dump_stacktraces
  File "/usr/lib64/python2.7/socket.py", line 476, in readline
  File "/usr/lib64/python2.7/httplib.py", line 620, in _read_chunked
  File "/usr/lib64/python2.7/httplib.py", line 578, in read
  File "/usr/lib/python2.7/site-packages/urllib3/response.py", line 203, in read
  File "/usr/lib/python2.7/site-packages/docker/client.py", line 247, in _stream_helper
  File "/usr/lib/python2.7/site-packages/atomic_reactor/util.py", line 297, in wait_for_command
  File "/usr/lib/python2.7/site-packages/atomic_reactor/plugins/build_docker_api.py", line 46, in run
  File "/usr/lib/python2.7/site-packages/atomic_reactor/plugin.py", line 239, in run
  File "/usr/lib/python2.7/site-packages/atomic_reactor/plugin.py", line 449, in run
  File "/usr/lib/python2.7/site-packages/atomic_reactor/inner.py", line 443, in build_docker_image
  File "/usr/lib/python2.7/site-packages/atomic_reactor/inner.py", line 546, in build_inside
  File "/usr/lib/python2.7/site-packages/atomic_reactor/cli/main.py", line 95, in cli_inside_build
  File "/usr/lib/python2.7/site-packages/atomic_reactor/cli/main.py", line 292, in run
  File "/usr/lib/python2.7/site-packages/atomic_reactor/cli/main.py", line 310, in run
  File "/usr/bin/atomic-reactor", line 11, in <module>
2018-10-24 08:51:34,561 - atomic_reactor.util - DEBUG - ---> d281798a9096
2018-10-24 08:51:34,569 - atomic_reactor.util - DEBUG - Removing intermediate container 55065d909a13
2018-10-24 08:51:34,569 - atomic_reactor.util - DEBUG - Step 4/5 : ADD Dockerfile-athos-ribeiro-hello-osbs-1.0-1 /root/buildinfo/Dockerfile-athos-ribeiro-hello-osbs-1.0-1
```

**Comment by MartinBasti** [comment]
IMHO signal handling should be enabled as soon as possible, it may be useful even for stuff out of building itself.

**Comment by lcarva** [comment]
>  Just note, if we have documentation about debugging OSBS, this should be IMO added there.

Good call! @athos-ribeiro wanna file something for that?

**Pull request merged**